File: src/tui/editor.ts
/**
 * Resolving which editor to launch and how to tell it which line to open.
 *
 * Kept separate from the spawn itself so the argument logic is testable
 * without starting a process.
 */

import path from 'node:path';

/** A spawnable editor invocation. */
export interface EditorCommand {
  command: string;
  args: string[];
}

/** Editors that accept `+LINE` before the path. */
const PLUS_LINE = new Set(['vim', 'nvim', 'vi', 'nano', 'gvim']);

/** Editors that accept `-g path:LINE`. */
const GOTO_FLAG = new Set(['code', 'code-insiders', 'cursor', 'codium']);

/**
 * Split an editor string into argv tokens.
 *
 * Handles whitespace-separated tokens and simple single/double quoting
 * (`code --wait`, `nvim -p`, `"path with spaces/editor"`). Not a full shell —
 * no escapes, expansion, or operators.
 */
export function parseEditorArgv(editor: string): string[] {
  const tokens: string[] = [];
  let current = '';
  let quote: '"' | "'" | null = null;

  for (let i = 0; i < editor.length; i++) {
    const ch = editor[i]!;
    if (quote) {
      if (ch === quote) {
        quote = null;
      } else {
        current += ch;
      }
      continue;
    }
    if (ch === '"' || ch === "'") {
      quote = ch;
      continue;
    }
    if (/\s/.test(ch)) {
      if (current) {
        tokens.push(current);
        current = '';
      }
      continue;
    }
    current += ch;
  }
  if (current) tokens.push(current);
  return tokens;
}

/**
 * The editor to launch: non-blank config `editor`, then `$EDITOR`, then
 * `$VISUAL`, then `vim`. Blank values are ignored so an exported-but-empty
 * variable or whitespace-only config does not win.
 */
export function resolveEditor(env: NodeJS.ProcessEnv, configEditor?: string): string {
  const fromConfig = configEditor?.trim();
  if (fromConfig) return fromConfig;
  const editor = env.EDITOR?.trim();
  if (editor) return editor;
  const visual = env.VISUAL?.trim();
  if (visual) return visual;
  return 'vim';
}

/**
 * Command and arguments to open `filePath`, at `line` when the editor is known
 * to support it. Unknown editors receive the path alone rather than a flag
 * they might treat as a filename.
 *
 * Multi-token values (`code --wait`, `nvim -p`) are split: the first token is
 * the executable, remaining tokens are fixed args prepended before line/path
 * args. PLUS_LINE / GOTO_FLAG matching uses the basename of the executable.
 */
export function editorCommand(editor: string, filePath: string, line?: number): EditorCommand {
  const argv = parseEditorArgv(editor);
  const command = argv[0] ?? editor;
  const fixedArgs = argv.slice(1);
  const name = path.basename(command);

  if (line === undefined) {
    return { command, args: [...fixedArgs, filePath] };
  }
  if (PLUS_LINE.has(name)) {
    return { command, args: [...fixedArgs, `+${line}`, filePath] };
  }
  if (GOTO_FLAG.has(name)) {
    return { command, args: [...fixedArgs, '-g', `${filePath}:${line}`] };
  }
  return { command, args: [...fixedArgs, filePath] };
}

/** GUI editors that return immediately and must not steal the TTY / remount Ink. */
const DETACHED_EDITORS = new Set(['code', 'code-insiders', 'cursor', 'codium', 'gvim']);

/**
 * True when `e` should spawn the editor without unmounting the TUI.
 *
 * Cursor / VS Code family (and gvim) are separate GUI processes. Their CLIs
 * typically exit as soon as the window or tab opens, so the old remount loop
 * rebuilt the whole tree immediately. TTY editors (vim, nvim, nano, and
 * unknown names) still use the blocking inherit path.
 *
 * `--wait` / `-w` does not change this. The GUI still does not need the TTY,
 * and staying mounted keeps fold, focus, and scroll.
 */
export function isDetachedEditor(editor: string): boolean {
  const argv = parseEditorArgv(editor);
  const command = argv[0] ?? editor;
  return DETACHED_EDITORS.has(path.basename(command));
}

File: crates/workspace-status/src/tui/editor.rs
//! Resolve `$EDITOR` / config.editor and build the spawn argv.

use std::path::Path;

/// Editors that accept `+LINE` before the path.
const PLUS_LINE: &[&str] = &["vim", "nvim", "vi", "nano", "gvim"];
/// Editors that accept `-g path:LINE`.
const GOTO_FLAG: &[&str] = &["code", "code-insiders", "cursor", "codium"];
/// GUI editors that must not steal the TTY.
const DETACHED: &[&str] = &["code", "code-insiders", "cursor", "codium", "gvim"];

/// Split an editor string into argv tokens. Simple quotes only.
pub fn parse_editor_argv(editor: &str) -> Vec<String> {
    let mut tokens = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    for ch in editor.chars() {
        if let Some(q) = quote {
            if ch == q {
                quote = None;
            } else {
                current.push(ch);
            }
            continue;
        }
        if ch == '"' || ch == '\'' {
            quote = Some(ch);
            continue;
        }
        if ch.is_whitespace() {
            if !current.is_empty() {
                tokens.push(std::mem::take(&mut current));
            }
            continue;
        }
        current.push(ch);
    }
    if !current.is_empty() {
        tokens.push(current);
    }
    tokens
}

/// Config `editor`, then `$EDITOR`, then `$VISUAL`, then `vim`.
pub fn resolve_editor(config_editor: Option<&str>, env_editor: Option<&str>, env_visual: Option<&str>) -> String {
    for candidate in [config_editor, env_editor, env_visual] {
        if let Some(raw) = candidate {
            let trimmed = raw.trim();
            if !trimmed.is_empty() {
                return trimmed.to_string();
            }
        }
    }
    "vim".into()
}

/// True when `e` should spawn without leaving the alternate screen.
pub fn is_detached_editor(editor: &str) -> bool {
    let argv = parse_editor_argv(editor);
    let command = argv.first().map(String::as_str).unwrap_or(editor);
    let name = Path::new(command)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(command);
    DETACHED.contains(&name)
}

/// Executable plus args to open `file_path` (optional 1-based `line`).
pub fn editor_command(editor: &str, file_path: &str, line: Option<u32>) -> (String, Vec<String>) {
    let argv = parse_editor_argv(editor);
    let command = argv.first().cloned().unwrap_or_else(|| editor.to_string());
    let mut args: Vec<String> = argv.into_iter().skip(1).collect();
    let name = Path::new(&command)
        .file_name()
        .and_then(|s| s.to_str())
        .unwrap_or(&command)
        .to_string();
    if let Some(line) = line {
        if PLUS_LINE.contains(&name.as_str()) {
            args.push(format!("+{line}"));
            args.push(file_path.into());
            return (command, args);
        }
        if GOTO_FLAG.contains(&name.as_str()) {
            args.push("-g".into());
            args.push(format!("{file_path}:{line}"));
            return (command, args);
        }
    }
    args.push(file_path.into());
    (command, args)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn resolve_prefers_config_then_editor() {
        assert_eq!(resolve_editor(Some("nvim"), Some("vim"), Some("vi")), "nvim");
        assert_eq!(resolve_editor(Some("  "), Some("vim"), None), "vim");
        assert_eq!(resolve_editor(None, None, None), "vim");
        assert_eq!(resolve_editor(None, None, Some("nano")), "nano");
    }

    #[test]
    fn parse_quoted_and_flags() {
        assert_eq!(
            parse_editor_argv(r#"cursor --wait"#),
            vec!["cursor", "--wait"]
        );
        assert_eq!(
            parse_editor_argv(r#""/path with spaces/vim" -p"#),
            vec!["/path with spaces/vim", "-p"]
        );
    }

    #[test]
    fn detached_gui_vs_tty() {
        assert!(is_detached_editor("cursor"));
        assert!(is_detached_editor("code --wait"));
        assert!(!is_detached_editor("vim"));
        assert!(!is_detached_editor("nvim -p"));
    }

    #[test]
    fn plus_line_for_vim() {
        let (cmd, args) = editor_command("vim", "README.md", Some(12));
        assert_eq!(cmd, "vim");
        assert_eq!(args, vec!["+12", "README.md"]);
    }
}
